com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/ShaderPass/DotsDeformation.hlsl

```diff
@@ -0,0 +1,66 @@
+#if defined(DOTS_INSTANCING_ON)
+struct DeformedVertexData
+{
+    float3 Position;
+    float3 Normal;
+    float3 Tangent;
+};
+
+uniform StructuredBuffer<DeformedVertexData> _DeformedMeshData;
+uniform StructuredBuffer<DeformedVertexData> _PreviousFrameDeformedMeshData;
+
+// Reads vertex data for compute skinned meshes in Hybdrid Renderer
+void FetchComputeVertexData(inout AttributesMesh input)
+{
+    // x,y = current and previous frame indices
+    // z = deformation check (0 = no deformation, 1 = has deformation)
+    // w = skinned motion vectors
+    const int4 deformProperty = asint(unity_DOTSDeformationParams);
+    const int doSkinning = deformProperty.z;
+    if (doSkinning == 1)
+    {
+        const uint streamIndex = _HybridDeformedVertexStreamIndex;
+        const uint startIndex = deformProperty[streamIndex];
+        const DeformedVertexData vertexData = _DeformedMeshData[startIndex + input.vertexID];
+
+        input.positionOS = vertexData.Position;
+#ifdef ATTRIBUTES_NEED_NORMAL
+        input.normalOS = vertexData.Normal;
+#endif
+#ifdef ATTRIBUTES_NEED_TANGENT
+        input.tangentOS = float4(vertexData.Tangent, 0);
+#endif
+    }
+}
+
+// Reads vertex position for compute skinned meshes in Hybdrid Renderer
+// and also previous frame position if skinned motion vectors are used
+void FetchComputeVertexPosition(inout float3 currPos, inout float3 prevPos, uint vertexID)
+{
+    // x,y = current and previous frame indices
+    // z = deformation check (0 = no deformation, 1 = has deformation)
+    // w = skinned motion vectors
+    const int4 deformProperty = asint(unity_DOTSDeformationParams);
+    const int computeSkin = deformProperty.z;
+    const uint streamIndex = _HybridDeformedVertexStreamIndex;
+    if (computeSkin == 1)
+    {
+        const uint currMeshStart = deformProperty[streamIndex]; 
+        const uint currStreamIndex = _HybridDeformedVertexStreamIndex;
+        currPos = _DeformedMeshData[currMeshStart + vertexID].Position;
+    }
+
+    const int skinMotionVec = deformProperty.w;
+    if (skinMotionVec == 1)
+    {
+        const uint prevStreamIndex = streamIndex ^ 1;
+        const int prevMeshStart = deformProperty[prevStreamIndex];
+
+        if(prevMeshStart == -1)
+            prevPos = _DeformedMeshData[prevMeshStart + vertexID].Position;
```

Comment:
Suggetsion / Remark. It could be more efficient to have a single StructuredBuffer<DeformedVertexData> with both the previous and current position and dealing with and offset instead of branching on buffer read like this which waste VGPR.

i.e

StructuredBuffer<DeformedVertexData> _FrameDeformedMeshData

int offset = (prevMeshStart == -1) ? 0 : nextStart;
offset +=  vertexID;
prevPos  = _FrameDeformedMeshData[offset].Position;

=> no branch, less sgpr/vgpr.

thought?

Reply:
hmm currently I am confuse by the code, it looks like you alrady do that. with prevMeshStart  and currMeshStart , so why is there a different _PreviousFrameDeformedMeshData structure buffer?  

sorry if my question is stupid I don't know the underlaying of DOTS skinning